### PR TITLE
feat: improve entries with include and recursive

### PR DESCRIPTION
**Pull request by TomChv**
Support include pattern.
Support recursive option.
Add tests

Replace and complete #5686 

Related to: #5685

### CODE REVIEW

File: core/integration/directory_test.go
Comment (jedevc) [inline core/integration/directory_test.go:981]: I'm not actually convinced this belongs in `Entries` (happy to be told I'm wrong though).

Today `Entries` returns the contents of a target directory - with these new options, it instead does something different - it tells you all the files that match a given glob. It feels like two different functionalities squashed into a single call depending on the parameters.

I wonder if we could instead add a new method, `Glob` (or similar)? This would just return a list of all the files that match the given glob - as a nice plus, we can also directly use `IncludePatterns` from upstream buildkit (though we still need to call `ReadDir` recursively, unless we want to upstream something like that into BuildKit).

File: core/directory.go
Comment (jedevc) [inline core/directory.go:268]: I don't understand why we don't just use `IncludePatterns` here?

It doesn't recurse, so we'd still need multiple calls to it - but the directories should still be returned.

Comment (jedevc) [reply core/directory.go:268]: We'd also need to update tests: in both tests `*.md` should be `**/*.md` to represent all markdown files in all subdirectories. `*.md` should only match top-level markdown files (in keeping with shell tradition).

File: core/integration/directory_test.go
Comment (TomChv) [reply core/integration/directory_test.go:981]: > Or we could always assume Recursive when Include is set - which would be functionally equivalent to the above, it's just a naming convention as to whether it's under Glob or under Entries.

`Recursive` and `Include` are 2 different option, if I want `*.md`, I do not expected a recursivity. However I can add a check to see if the glob contains `**` and then activate `Recursive` in that way, this could works.

> IncludePatterns from upstream buildkit (though we still need to call ReadDir recursively, unless we want to upstream something like that into BuildKit).

See https://github.com/dagger/dagger/pull/5824#discussion_r1337617427

> Today Entries returns the contents of a target directory - with these new options, it instead does something different - it tells you all the files that match a given glob. It feels like two different functionalities squashed into a single call depending on the parameters.

I'm not sure it's right to put it in another method, this would mostly lead to code duplication, if an user want to get the content of the target directory, he can also expect to get subdirectories as mentioned in #5685.

WDYT @vito @helderco @gerhard

File: core/directory.go
Comment (jedevc) [reply core/directory.go:268]: IncludePatterns only returns directories that contain files that are matched by the glob. e.g. for a pattern `**/*.md` a directory `dir` will only be returned if it (or a child of it) contains a `.md` file. Is this what you're referring to?

IMO `*.md` should never match `foo/bar/*.md`. You should need the pattern `**/*.md` to catch it. Essentially, the glob should probably behave exactly as it does in bash/zsh/all common implementations of sh.

File: core/integration/directory_test.go
Comment (vito) [reply core/integration/directory_test.go:981]: I agree with @jedevc; I don't think we need `Recursive`. `*.md` should always match the top level, and `**/*.md` should be how you specify recursive globs.

My suggestion API wise is this:

```graphql
extend type Directory {
  entries(include:[String!], exclude: [String!]): [String!]!
}
```

This would match the various include/exclude filter APIs we have everywhere else.

I can see the appeal of having a separate `Glob` API instead, but having separate include/exclude filters implies both are optional, so you might as well just put it on `Entries` IMO. (Unless there's some sort of built-in 'negated glob' syntax I'm not aware of.)

File: core/directory.go
Comment (TomChv) [reply core/directory.go]: You're right! Fixed in [4364df0](https://github.com/dagger/dagger/pull/5824/commits/4364df05a7849233a953d8c94155e72a985a08cc) to use `IncludePattern`

File: core/integration/directory_test.go
Comment (TomChv) [reply core/integration/directory_test.go]: `IncludePattern` Only expect a string, I'm not sure we can support multiple include / exclude.
In our case excluse will have to be done manually too, I would prefer supporting that in a further PR

I added your suggestion about `Recursive`deletion in [4364df0](https://github.com/dagger/dagger/pull/5824/commits/4364df05a7849233a953d8c94155e72a985a08cc)

**Comment by gerhard** [comment]
I will review this if I can make time, but I would appreciate your eyes on this @helderco 👀

**Comment by TomChv** [comment]
@jedevc Would be cool to get your review too again on that one 🚀

File: core/directory.go
Comment (jedevc) [inline core/directory.go:268]: I'm not sure about this - it would be nice to avoid introspecting this string. But I think it works for now. If we had multiple include patterns, this would the logic for this would be quite trick I think - one pattern could be recursive, while the other wouldn't be.

Wdyt about instead *always* enabling `recursive = true` if `include` isn't empty? I don't see a use case where `include` is set without a recursive path and `src` is also set.

This is sort of what I meant about having a separate `Glob` method - it's really annoying to work out when the recursive case should occur. I don't want to block, but still think that might be the way to go.

Comment (jedevc) [reply core/directory.go:268]: Also the different path manipulation based on `recursive` is quite confusing. I think there's such differences in the return values (they're really not interchangeable), that I think that adds more weight to splitting it out to another method.

File: core/integration/directory_test.go
Comment (jedevc) [reply core/integration/directory_test.go]: > In our case excluse will have to be done manually too, I would prefer supporting that in a further PR

Agreed about future PR for now. However - we should instead upstream the support for this to buildkit (should be a relatively simple fix that we can merge quickly, happy to pair on it :tada:), instead of adding custom logic.

File: core/directory.go
Comment (TomChv) [reply core/directory.go:268]: If you do `include: *.md`, you only expect files that match the pattern in the current directory

Comment (jedevc) [reply core/directory.go:268]: If `include: *.md`, then `ReadDir` only returns files in the current directory, and never anything below.

Comment (TomChv) [reply core/directory.go:268]: If I do that, the issue is that output will include the `src` if it's set where we only expect it to be print for sub directories

**Comment by TomChv** [comment]
I just had a call with @jedevc to talk about this issue and how we should implement it.

Few thing will chance before we merge that PR:
- `Entries` will not support `includePattern` option, we will keep it as it is for now.
- We will create a new method `Glob` that will do almost the same as entry but will supports an option `pattern` that will handle the pattern matching and possible recursivity.
- The current way we handle recursivity isn't optimal because we call `ReadDir` multiple times, meaning that we mount and unmount the directory each time, this behaviour could lead to poor perfomrance. We could actually fix that by improving the Buldkit ReadDir function to take care of the globing in one mount/dismount operation. However, since it requires to copy/paste some buildkit function, we'll hold this for later if an user require more performance when call `Glob`.
- We will use  [`moby/patternmatcher`](https://godocs.io/github.com/moby/patternmatcher) to check for pattern if needed (I might actually don't need it)

We came up with that change because the support of `includePattern` on `Entries` actually changes return values that a user will expect (specially when subdirectory path are appended to the file), to make the DX better, it's easier to simply add a new method even if it lead to some code duplication.
Specially since `path` and `includePattern` can do the same thing at some point, for instance: `path: ./a/b, includePattern: *md` == `includePattern: ./a/b/*.md`, to avoid confusion we will now have 2 methods.

/cc @vito @sipsma @aluzzardi for context, I'm waiting for one of your approval to implement this solution.

**Comment by TomChv** [comment]
@jedevc Ready for a review :D

File: core/directory.go
Comment (TomChv) [reply core/directory.go]: Fixed in https://github.com/dagger/dagger/pull/5824/commits/dd6f9a1df32704cdaf0886ffe00af7bd9f047719

File: core/integration/directory_test.go
Comment (TomChv) [reply core/integration/directory_test.go:931]: Fixed in https://github.com/dagger/dagger/pull/5824/commits/dd6f9a1df32704cdaf0886ffe00af7bd9f047719

**Review by jedevc — APPROVED** [review]

**Review by sipsma — APPROVED** [review]

**Pull request merged**
